packages/contracts-bedrock/test/scripts/L2Genesis.t.sol

```diff
@@ -0,0 +1,242 @@
+import { Executables } from "scripts/Executables.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { L2GenesisHelpers } from "scripts/libraries/L2GenesisHelpers.sol";
+import { L2Genesis } from "scripts/L2Genesis.s.sol";
+
+/// @notice Reads a `genesis-l2.json` file, parses the `alloc`s, and runs assertions
+///         against each alloc depending on whether it's a precompile, predeploy proxy,
+///         or predeploy implementation.
+contract L2Genesis_Test is Test, Artifacts {
+    DeployConfig public constant cfg =
+        DeployConfig(address(uint160(uint256(keccak256(abi.encode("optimism.deployconfig"))))));
+
+    string internal genesisPath;
+    L2Genesis internal l2Genesis = new L2Genesis();
+
+    struct StorageData {
+        bytes32 key;
+        bytes32 value;
+    }
+
+    /// @notice `balance` and `nonce` are being parsed as `bytes` even though their JSON representations are hex
+    /// strings.
+    ///         This is because Foundry has a limitation around parsing strings as numbers when using `vm.parseJson`,
+    ///         and because we're using `abi.decode` to convert the JSON string, we can't use coersion (i.e.
+    /// `vm.parseJsonUint`)
+    ///         to tell Foundry that the strings are numbers. So instead we treat them as `byte` strings and parse as
+    ///         `uint`s when needed. Additional context: https://github.com/foundry-rs/foundry/issues/3754
+    struct Alloc {
+        address addr;
+        bytes balance;
+        bytes code;
+        bytes nonce;
+        StorageData[] storageData;
+    }
+
+    /// @notice Checks if deploy config file exists at `deployConfigPath`, if it doesn't
+    ///         reverts with warning to run `make devnet-allocs` at monorepo root.
+    ///         Afterwards, checks if genesis file exists at `genesisPath`, and if not
+    ///         runs `L2Genesis.s.sol` script to generate it.
+    function setUp() public override {
+        super.setUp();
+        Artifacts.setUp();
+
+        string memory deployConfigPath = string.concat(vm.projectRoot(), "/deploy-config/", deploymentContext, ".json");
+        if (!vm.exists(deployConfigPath)) {
+            revert(
+                string.concat(
+                    "Did not find deploy config at: ",
+                    deployConfigPath
+                )
+            );
+        }
+
+        vm.etch(address(cfg), vm.getDeployedCode("DeployConfig.s.sol:DeployConfig"));
+        vm.label(address(cfg), "DeployConfig");
+        vm.allowCheatcodes(address(cfg));
+        cfg.read(deployConfigPath);
+
+        genesisPath = string.concat(vm.projectRoot(), "/deployments/", deploymentContext, "/genesis-l2.json");
+        if (!vm.exists(genesisPath)) {
+            string[] memory commands = new string[](3);
+            commands[0] = "bash";
+            commands[1] = "-c";
+            commands[2] = string.concat(
+                "DEPLOYMENT_CONTEXT=",
+                deploymentContext,
```

Comment:
We don't need this anymore? The `L2Genesis` contract is in context

Reply:
```solidity
function setUp() public {
    vm.etch(address(l2Genesis), vm.getDeployedCode("L2Genesis.s.sol:L2Genesis"));
    l2Genesis.setUp();

    genesisPath = string.concat(vm.projectRoot(), "/deployments/", l2Genesis.getDeploymentContext(), "/genesis-l2.json");
    if (!vm.exists(genesisPath)) {
        console.log("No L2 genesis found at %s, generating a new one", genesisPath);
        l2Genesis.run();
    }
}
```

With this code, if I set`l2Genesis` to the actual address of `L2Genesis.s.sol`

```solidity
L2Genesis internal l2Genesis = L2Genesis(0x7Fa9385be102AC3EAC297483dD6233d62B3E1496);
```

the `vm.etch` causes a revert.

Removing the `vm.etch` causes the `l2Genesis.setUp();` to revert

```
 setUp failed reason=EvmError: Revert contract=0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496
```

and `vm.etch`ing `L2Genesis` to a random address causes the tests to revert because the actual `L2Gensis.s.sol` is included in the genesis outfile:

```
    └─ ← revert: Unknown alloc: 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 3.84s

Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/scripts/L2Genesis.t.sol:L2Genesis_Test
[FAIL. Reason: revert: Unknown alloc: 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496] test_allocs() (gas: 4713561890)
```

So unless I'm missing something, I don't think we can `run()` `L2Genesis.s.sol` from within the test script

Reply:
This is [related](https://github.com/ethereum-optimism/optimism/pull/9410#discussion_r1483944617)